# [Transcriptomics]{style="color:purple;"}

## Author: Carolyn Hanrahan

### Affiliation: University of Vermont, RSENR

### E-mail contact: [carolyn.hanrahan\@uvm.edu](mailto:carolyn.hanrahan@uvm.edu)

### Start Date: 10/16/2023

### End Date: TBD

### Project Descriptions: This notebook markdown file will document my workflow of the bioinformatics of the Transcriptomics section of Ecological Genomics, fall 2023.

# Table of Contents:

-   [Entry 1: 2023-10-16](#id-section1)
-   [Entry 2: 2023-10-18](#id-section2)
-   [Entry 3: 2023-10-23](#id-section3)
-   [Entry 4: 2023-10-25](#id-section4)
-   [Entry 5: 2023-09-25](#id-section5)
-   [Entry 6: 2023-09-27](#id-section6)
-   [Entry 7: 2023-10-02](#id-section7)
-   [Entry 8: 2023-10-04](#id-section8)

<div id='id-section1'/>

### Entry 1: 2023-10-09

-   Transcriptomics unit; reading paper on Colorado Potato Beetle and insecticide resistance.

-   Coding goals for the day:

    -   Review Acartia hudsonica ecology and biogeography and the experimental evolution/transcriptomics experimental design.

    -   Develop questions that can be addressed and hypotheses that can be tested with the A. hudsonica experiment.

    -   Understand the general work flow or "pipeline" for processing and analyzing RNAseq data.

    -   Visualize and interpret the quality of our Illumina data.

    -   Assess our previously assembled de novo transcriptome assembly using Trinity. Start mapping reads and quantifying abundance simultaneously using Salmon.

## Entry 2: 2023-10-11:

-   Transcriptomics DAY 2.
-   Copepod study
    -   effect of environment of Genetic Evolution
    -   signature of physiological adaptation in F2, F4, F11...
    -   does the gene response in a given environment depend on the generation? We will conduct a statistical test for interaction between environmental conditions and generation.
    -   Interaction between environment and generation
-   Today we are...
    -   cleaning the raw sequence data
    -   generating and annotating a de novo reference transcriptome

### Entry 3: 2023-10-16.

-   Transcriptomics Day 3.

-   Goals for the day:

    -   Review the quality of our Illumina data from our google sheet.
    -   Assess our previously assembled de novo transcriptome assembly using scripts provided through Trinity.
    -   Map reads and quantify abundance simultaneously using Salmon.
    -   Prepare data for import into DESeq2.
    -   Start analyzing the gene expression data using DESeq2

<div id='id-section2'/>

### Entry 4: 2023-10-18.

-   Transcriptomics Day 4.

-   Goals for the day:

    -   Assemble the quant.sf files generated by salmon into a matrix to import into DESeq2.
    -   Start analyzing the gene expression data using DESeq2

<div id='id-section3'/>

### Entry 5: 2023-10-23.

-   Transcriptomics Day 5.

-   Goals for the day:

    -   Review results from filtering of the de novo assembly and transfer the new gene expression matrix.
    -   Explore patterns in the gene expression data using DESeq2: explore data, plot PCA, test for DGE, make a Venn diagram.

-   see `transcriptomics_day4.R` file for scripts from class (added on from day 4)

<div id='id-section4'/>

### Entry 6: 2023-10-25

-   Transcriptomics day 6

-   We downloaded many R packages. We did a weighted gene correlation network analysis to look for clusters across transcripts. These transcripts are defined as "modules".

-   worked with `transcriptomics_day6.R` script.

-   Goals for today:

    -   Transfer a data file that includes additional trait information.

    -   Work step by step through correlation (or association) analyses using WGCNA: install packages, import and filter the data, look for outliers, normalize, construct networks, explore module eigengenes, associate module eigengenes with traits, and visualize the data along the way!

### Entry 7: 2023-10-30

-   Transcriptomics Day 7

-   review whale cancer paper.

-   Goals for today:

    -   Discuss "what's missing" in hands-on coding sessions - to better set you up for the future (so you're not surprised)

    -   Continue working through the correlation (or association) analyses using WGCNA (weighted gene co-expression network analysis): associate module eigengenes with traits and visualize the data!

    -   Perform Gene Ontology (GO) functional enrichment analyses using GOMWU with our DESeq2 results.

What's missing in our in-class coding sessions...

-   Choosing which programs to use, installing the programs for use, troubleshooting code, moving from one file type to the next, exploring parameter space. ChatGPT and Google and people can be helpful!

### Entry 8: 2023-11-01

-   Transcriptomics Day 8

-   Goals for today:

    -   Review Homework Assignment #2.

    -   Perform Gene Ontology (GO) functional enrichment analyses using GOMWU with our DESeq2 results.

### Entry 9: 2023-10-02

### Entry 8: 2023-10-04

### Entry 9: 2023-10-09

### Entry 10: 2023-10-11:
